Amazon.Ion.ObjectMapper/IonSerializer.cs

```diff
@@ -161,97 +163,79 @@ public void Serialize<T>(Stream stream, T item)
 
         public void Serialize<T>(IIonWriter writer, T item)
         {
+            currentDepth++;
+            if (currentDepth >= options.MaxDepth)
+            {
+                currentDepth = 0;
+                throw new NotSupportedException($"Can not serialize further as max object tree depth {options.MaxDepth} is reached");
+            }
+
             if (item == null)

```

Comment:
Not specifically related to this PR, but the if/else blocks here could be replaced with a dictionary keyed on type.

Reply:
Could you elaborate a bit further about this?

Reply:
Let's talk offline as it's an improvement that is unrelated to this change in general.